### PR TITLE
checker: require or block for sumtype map

### PR DESCRIPTION
**Pull request by danieldaeschle**
Fixes https://github.com/vlang/v/issues/10333

### CODE REVIEW

**Comment by danieldaeschle** [comment]
why is this failing?

File: vlib/v/compiler_errors_test.v [hunk 1]
```diff
@@ -290,7 +290,7 @@ fn (mut task TaskDescription) execute() {
 	task.expected_out_path = expected_out_path
 	task.cli_cmd = cli_cmd
 	if should_autofix && !os.exists(expected_out_path) {
-		os.write_file(expected_out_path, '') or { panic(err) }
+		os.create(expected_out_path) or { panic(err) }
```
Comment (spytheman) [inline vlib/v/compiler_errors_test.v:293]: why is that change needed?

Comment (spytheman) [reply vlib/v/compiler_errors_test.v:293]: `os.write_file` opens a file and writes to it, then closes the file.
`os.create` only creates a file, but neither writes to it, nor closes it.

File: vlib/v/ast/scope.v [hunk 2]
```diff
@@ -45,7 +45,7 @@ pub fn (s &Scope) find_with_scope(name string) ?(ScopeObject, &Scope) {
 	mut sc := s
 	for {
 		if name in sc.objects {
-			return sc.objects[name], sc
+			return unsafe { sc.objects[name] }, sc
```
Comment (medvednikov) [inline vlib/v/ast/scope.v:48]: why unsafe here?

File: vlib/v/compiler_errors_test.v (hunk 1 above)
Comment (danieldaeschle) [reply vlib/v/compiler_errors_test.v:293]: Because it should create the file if it doesn't exist. The previous code didn't work and the expected behavior was creating anyway.

File: vlib/v/ast/scope.v (hunk 2 above)
Comment (danieldaeschle) [reply vlib/v/ast/scope.v:48]: Accessing maps with sumtype value should either be used with or{} or wrapped with unsafe.

File: vlib/v/compiler_errors_test.v (hunk 1 above)
Comment (medvednikov) [reply vlib/v/compiler_errors_test.v:293]: I see that makes sense. Maps return zero values, and there's no zero value for sumtypes, just `nil`.

File: vlib/v/ast/scope.v (hunk 2 above)
Comment (medvednikov) [reply vlib/v/ast/scope.v:48]: I see that makes sense. Maps return zero values, and there's no zero value for sumtypes, just nil.

Comment (medvednikov) [reply vlib/v/ast/scope.v:48]: I'm not a fan of unsafe though. Perhaps sum types should have default values, kind of like 0 is the default value for ints.

Comment (medvednikov) [reply vlib/v/ast/scope.v:48]: So perhaps the first value is the default value?

File: vlib/v/compiler_errors_test.v (hunk 1 above)
Comment (spytheman) [reply vlib/v/compiler_errors_test.v:293]: How is it not creating the file?
It is executing inside a check that the file does not yet exist.

Comment (danieldaeschle) [reply vlib/v/compiler_errors_test.v:293]: Try it out yourself. It paniced for me.

File: vlib/v/ast/scope.v (hunk 2 above)
Comment (danieldaeschle) [reply vlib/v/ast/scope.v:48]: That's a good idea. I think we need to discuss that first on discord.

Comment (danieldaeschle) [reply vlib/v/ast/scope.v:48]: Since the default value for sum types is no longer a solution what do you want instead?

File: vlib/v/compiler_errors_test.v (hunk 1 above)
Comment (danieldaeschle) [reply vlib/v/compiler_errors_test.v:293]: I don't know. Is there a difference between create and write_file with empty content?

**Comment by danieldaeschle** [comment]
Any update on this? @medvednikov

**Comment by medvednikov** [comment]
@danieldaeschle yeah this can be merged, sorry for the delay

can you fix the one conflict please?

**Comment by medvednikov** [comment]
Hm it's a big change. Can you make it a warning for now?

I'll change it to an error in 1-2 weeks.

**Comment by danieldaeschle** [comment]
Ok, will do it.

File: vlib/v/checker/checker.v [hunk 3]
```diff
@@ -7425,6 +7425,12 @@ pub fn (mut c Checker) index_expr(mut node ast.IndexExpr) ast.Type {
 				err := c.expected_msg(index_type, info.key_type)
 				c.error('invalid key: $err', node.pos)
 			}
+			value_sym := c.table.get_type_symbol(info.value_type)
+			if !node.is_setter && value_sym.kind == .sum_type && node.or_expr.kind == .absent
+				&& !c.inside_unsafe {
+				c.error('`or {}` block required when indexing a map with sum type value',
```
Comment (danieldaeschle) [inline vlib/v/checker/checker.v:7431]:
```suggestion
				c.warn('`or {}` block required when indexing a map with sum type value',
```

File: vlib/v/checker/tests/require_or_block_sumtype_map.err.out [hunk 4]
```diff
@@ -0,0 +1,6 @@
+vlib/v/checker/tests/require_or_block_sumtype_map.err.vv:5:8: error: `or {}` block required when indexing a map with sum type value
```
Comment (danieldaeschle) [inline vlib/v/checker/tests/require_or_block_sumtype_map.err.out:1]:
```suggestion
vlib/v/checker/tests/require_or_block_sumtype_map.err.vv:5:8: warning: `or {}` block required when indexing a map with sum type value
```

File: vlib/v/checker/tests/require_or_block_sumtype_map.err.out [hunk 5]
```diff
@@ -0,0 +1,6 @@
+vlib/v/checker/tests/require_or_block_sumtype_map.err.vv:5:8: warning: `or {}` block required when indexing a map with sum type value
```
Comment (danieldaeschle) [inline vlib/v/checker/tests/require_or_block_sumtype_map.err.out:1]:
```suggestion
vlib/v/checker/tests/require_or_block_sumtype_map.err.vv:5:8: error: `or {}` block required when indexing a map with sum type value
```

**Comment by danieldaeschle** [comment]
I don't know why vinix is failig.

**Comment by medvednikov** [comment]
Merged, thanks.
